### DISCUSSION TITLE
@Controller 와 @RestController 의 차이

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#139] [Spring MVC]
.

**Comment by jimi567** [comment]
둘 모두 Spring MVC에서 컨트롤러를 정의하는데에 사용되는 어노테이션이지만 다음과 같은 차이가 있습니다.  

### @Controller

+ 역할 : Spring MVC에서 웹 페이지를 반환하는 컨트롤러로 보편적으로 사용됩니다.
+ 작동 방식 : 메서드가 문자열을 반환하면, 해당 문자열을 뷰 이름으로 간주하고 뷰 리졸버가 이를 처리하여 클라이언트에 HTML 페이지가 반환 됩니다.

### @RestController 

+ 역할 : Restful API 컨트롤러로 사용됩니다.
+ 작동 방식 : @Controller 와 @ResponseBody를 결합한 어노테이션으로 메서드가 반환하는 값은 뷰 리졸버를 거치지 않고 HTTP 응답 본문에 적절한 Content-type로 직렬화 되어 쓰여집니다.  

--> 결론 

@RestController는 @Controller + @ResponseBody가 결합된 형태로 API 용도의 컨트롤러에 주로 사용되고, 단순 페이지만 내려주는 view Controller의 경우 @Controller 어노테이션을 사용하는게 일반적입니다.  가독성 측면에도 restful api에서 HTTP 응답을 하는 컨트롤러의 경우 @RestController을 사용하여 해당 컨트롤러 클래스가 API 용도의 컨트롤러 클래스임을 명시해줄 수 있습니다.

**Comment by GIVEN53** [comment]
이 두 어노테이션의 주요 차이점은 HTTP 응답의 처리 방식에 있습니다.

### `@Controller`
주로 뷰(View)를 반환하는 컨트롤러를 정의할 때 사용됩니다.
메서드가 반환하는 값은 뷰 리졸버(View Resolver)에 의해 해석되어 JSP, Thymeleaf 등과 같은 템플릿 엔진을 통해 HTML을 생성합니다.

### `@RestController`
주로 RESTful 웹 서비스 API를 정의할 때 사용됩니다.
메서드가 반환하는 값은 자동으로 JSON 또는 XML 형식으로 변환되어 HTTP 응답 본문에 포함됩니다. 이는 `@Controller`와 `@ResponseBody`의 결합된 형태입니다.
